### PR TITLE
Use the stateful overload of AsyncLock in some places.

### PR DESCRIPTION
**Pull request by danielcweber**
Save allocation of lots of closures and allow delegate caching!

### CODE REVIEW

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Buffer.cs [hunk 1]
```diff
@@ -155,7 +155,7 @@ protected override void Dispose(bool disposing)
                     if (disposing)
                     {
                         Disposable.TryDispose(ref _timerSerial);
-                    }
+                }
```
Comment (akarnokd) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/Buffer.cs]: Misaligned `}`

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Buffer.cs [hunk 2]
```diff
@@ -305,7 +305,7 @@ protected override void Dispose(bool disposing)
                     if (disposing)
                     {
                         Disposable.TryDispose(ref _periodicDisposable);
-                    }
+                }
```
Comment (akarnokd) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/Buffer.cs]: Misaligned `}`

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Buffer.cs [hunk 3]
```diff
@@ -399,7 +399,7 @@ protected override void Dispose(bool disposing)
                     if (disposing)
                     {
                         Disposable.TryDispose(ref _timerSerial);
-                    }
+                }
```
Comment (akarnokd) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/Buffer.cs]: Misaligned `}`

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Buffer.cs [hunk 4]
```diff
@@ -520,15 +520,15 @@ public override void Run(IObservable<TSource> source)
...
+                    _bufferGate.Wait(this, @this => @this.CreateBufferClose());
                 }
 
                 protected override void Dispose(bool disposing)
                 {
                     if (disposing)
                     {
                         Disposable.TryDispose(ref _bufferClosingSerialDisposable);
-                    }
+                }
```
Comment (akarnokd) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/Buffer.cs]: Misaligned `}`

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Buffer.cs [hunk 5]
```diff
@@ -663,7 +663,7 @@ protected override void Dispose(bool disposing)
                     if (disposing)
                     {
                         Disposable.TryDispose(ref _boundariesDisposable);
-                    }
+                }
```
Comment (akarnokd) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/Buffer.cs]: Misaligned `}`

Comment (danielcweber) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/Buffer.cs]: Unclear how that happened....fixed.

**Pull request merged**